quickwit-indexing/src/source/push_api_source.rs

```diff
@@ -0,0 +1,322 @@
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use async_trait::async_trait;
+use quickwit_actors::{ActorExitStatus, Mailbox};
+use quickwit_config::PushApiSourceParams;
+use quickwit_metastore::checkpoint::{CheckpointDelta, PartitionId, Position, SourceCheckpoint};
+use quickwit_proto::push_api::{FetchRequest, FetchResponse};
+use quickwit_pushapi::{get_push_api_service, iter_doc_payloads, PushApiService};
+use serde::Serialize;
+
+use super::{Source, SourceContext, TypedSourceFactory};
+use crate::actors::Indexer;
+use crate::models::RawDocBatch;
+
+/// Cut a new batch as soon as we have read BATCH_NUM_BYTES_THRESHOLD.
+const BATCH_NUM_BYTES_THRESHOLD: u64 = 500_000u64; // 0.5MB
+
+#[derive(Default, Clone, Debug, Eq, PartialEq, Serialize)]
+pub struct PushApiSourceCounters {
+    pub previous_offset: u64,
+    pub current_offset: u64,
+    pub num_docs_processed: u64,
+}
+
+pub struct PushApiSource {
+    params: PushApiSourceParams,
+    push_api_mailbox: Mailbox<PushApiService>,
+    counters: PushApiSourceCounters,
+}
+
+impl PushApiSource {
+    pub fn make(
+        params: PushApiSourceParams,
+        push_api_mailbox: Mailbox<PushApiService>,
+        checkpoint: SourceCheckpoint,
+    ) -> anyhow::Result<Self> {
+        let partition_id = PartitionId::from(params.index_id.clone());
+        let offset = if let Some(Position::Offset(offset_str)) =
+            checkpoint.position_for_partition(&partition_id).cloned()
+        {
+            offset_str.parse::<u64>()?
+        } else {
+            0
+        };
+
+        let push_api_source = PushApiSource {
+            params,
+            push_api_mailbox,
+            counters: PushApiSourceCounters {
+                previous_offset: offset,
+                current_offset: offset,
+                num_docs_processed: 0,
+            },
+        };
+        Ok(push_api_source)
+    }
+}
+
+#[async_trait]
+impl Source for PushApiSource {
+    async fn emit_batches(
+        &mut self,
+        batch_sink: &Mailbox<Indexer>,
+        ctx: &SourceContext,
+    ) -> Result<(), ActorExitStatus> {
+        let start_after = if self.counters.current_offset == 0 {
+            None
+        } else {
+            Some(self.counters.current_offset)
+        };
+
+        let fetch_req = FetchRequest {
+            index_id: self.params.index_id.clone(),
+            start_after,
+            num_bytes_limit: self
+                .params
+                .batch_num_bytes_threshold
+                .or(Some(BATCH_NUM_BYTES_THRESHOLD)),
+        };
+        let fetch_resp = self
+            .push_api_mailbox
+            .ask_for_res(fetch_req)
+            .await
+            .map_err(|err| anyhow::anyhow!(err.to_string()))?;
+
+        let FetchResponse {
+            first_position: first_position_opt,
+            doc_batch: doc_batch_opt,
+        } = fetch_resp;
+
+        if first_position_opt.is_none() {
+            return Ok(());
```

Comment:
@fulmicoton 
should we return `ActorExitStatus::Success`? 
This will exit this SourceActor and we currently have no way of recreating the SourceActor when new docs arrive.

Reply:
This is related to #1229.

For the moment, can you write something there before finishing this ticket maybe?
My 50c is for this iteration we don't think to be simple. 

In other words, the actor does not return Success if it reaches the end.
It keeps polling for new message.

We need however a way to sleep for 1s or so.
This probably means tweaking the SourceActor a little bit.

One improvement suggested by @trinity-1686a was to have actors return a Duration to inform the SourceActor how long it should sleep.

In the future, we can be smarter and have the push API source actor sleep properly.